include/aspect/material_model/reaction_model/tian2019_solubility.h

```diff
@@ -0,0 +1,159 @@
+                         const unsigned int porosity_idx,
+                         unsigned int q) const;
+
+          /**
+           * Compute the maximum allowed bound water content at the input
+           * pressure and temperature conditions. This is used to determine
+           * how free water interacts with the solid phase.
+           * @param in Object that contains the current conditions.
+           * @param q unsigned int from 0-3 indexing which rock phase the equilbrium
+           * bound water content is being calculated for
+           */
+          std::vector<double> tian_equilibrium_bound_water_content(const MaterialModel::MaterialModelInputs<dim> &in,
+                                                                   unsigned int q) const;
+
+        private:
+          /**
+           * Parameters for the solubility model of Tian et al., 2019.
+           */
+
+          /**
+           * The maximum water content for each of the 4 rock types in the tian approximation
+           * method. These are important for keeping the polynomial bounded within reasonable
+           * values.
+           */
+          double tian_max_peridotite_water;
+          double tian_max_gabbro_water;
+          double tian_max_MORB_water;
+          double tian_max_sediment_water;
+
+          /**
+           *
+           * The following coefficients are taken from a publication from Tian et al., 2019, and can be found
+           * in Table 3 (Gabbro), Table B1 (MORB), Table B2 (Sediments) and Table B3 (peridotite).
+           * LR refers to the effective enthalpy change for devolatilization reactions,
+           * csat is the saturated mass fraction of water in the solid, and Td is the
+           * onset temperature of devolatilization for water.
+           */
+          std::vector<double> LR_peridotite_poly_coeffs {-19.0609, 168.983, -630.032, 1281.84, -1543.14, 1111.88, -459.142, 95.4143, 1.97246};
+          std::vector<double> csat_peridotite_poly_coeffs {0.00115628, 2.42179};
+          std::vector<double> Td_peridotite_poly_coeffs {-15.4627, 94.9716, 636.603};
+
+          std::vector<double> LR_gabbro_poly_coeffs {-1.81745, 7.67198, -10.8507, 5.09329, 8.14519};
+          std::vector<double> csat_gabbro_poly_coeffs {-0.0176673, 0.0893044, 1.52732};
+          std::vector<double> Td_gabbro_poly_coeffs {-1.72277, 20.5898, 637.517};
+
+          std::vector<double> LR_MORB_poly_coeffs {-1.78177, 7.50871, -10.4840, 5.19725, 7.96365};
+          std::vector<double> csat_MORB_poly_coeffs {0.0102725, -0.115390, 0.324452, 1.41588};
+          std::vector<double> Td_MORB_poly_coeffs {-3.81280, 22.7809, 638.049};
+
+          std::vector<double> LR_sediment_poly_coeffs {-2.03283, 10.8186, -21.2119, 18.3351, -6.48711, 8.32459};
+          std::vector<double> csat_sediment_poly_coeffs {-0.150662, 0.301807, 1.01867};
+          std::vector<double> Td_sediment_poly_coeffs {2.83277, -24.7593, 85.9090, 524.898};
```

Comment:
All of these numbers are fixed and will never change during the program runtime. You can tell this to the compiler, which will then optimize the code further and also prevent you from accidentally changing the values by changing their type from `std::vector<double>` to `constexpr std::array<double,N>` where N is the number of values in this array. Initialization and access to these arrays can remain unchanged.